src/System.Text.Json/src/System/Text/Json/JsonParseObject.cs

```diff
@@ -0,0 +1,350 @@
+            {
+                int location = BitConverter.ToInt32(_buffer, _start);
+                if (_buffer[location - 1] == '"' && _buffer[location + 4] == '"')
+                {
+                    return true;
+                }
+                return (_buffer[location] != 'n' || _buffer[location + 1] != 'u' || _buffer[location + 2] != 'l' || _buffer[location + 3] != 'l');
+            }
+        }
+
+        public JsonParseObject this[string index]
+        {
+            get
+            {
+                int length = BitConverter.ToInt32(_buffer, _start + 4);
+                var typeCode = _buffer[_start + 8];
+
+                if (length == 0)
+                {
+                    throw new KeyNotFoundException();
+                }
+
+                if (typeCode != (byte)JsonParser.JsonValueType.Object)
+                {
+                    throw new NullReferenceException();
+                }
+
+                for (int i = _start + RowSize; i <= _end; i += RowSize)
+                {
+                    length = BitConverter.ToInt32(_buffer, i + 4);
+                    typeCode = _buffer[i + 8];
+
+                    if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+                    {
+                        i += length * RowSize;
+                        continue;
+                    }
+
+                    int location = BitConverter.ToInt32(_buffer, i);
+                    if (isEqual(index, _buffer, location, length))
+                    {
+                        int newStart = i + RowSize;
+                        int newEnd = newStart + RowSize;
+
+                        typeCode = _buffer[newStart + 8];
+
+                        if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+                        {
+                            length = BitConverter.ToInt32(_buffer, newStart + 4);
+                            newEnd = newEnd + RowSize * length;
+                        }
+
+                        return new JsonParseObject(_buffer, newStart, newEnd);
+                    }
+
+                    typeCode = _buffer[i + RowSize + 8];
+
+                    if (typeCode != (byte)JsonParser.JsonValueType.Object && typeCode != (byte)JsonParser.JsonValueType.Array)
+                    {
+                        i += RowSize;
+                    }
+                }
+
+                throw new KeyNotFoundException();
+
+            }
+        }
+
+        private bool isEqual(string str, byte[] buffer, int location, int length)
+        {
+            if (str.Length != length) return false;
+            for (int i = 0; i < length; i++)
+            {
+                if (str[i] != buffer[location + i])
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public JsonParseObject this[int index]
```

Comment:
I don't understand. I think the name JsonParseObject is causing some confusion. Integer indexer  only makes sense for arrays just like string indexer makes sense for objects. This class is only there to encapsulate the DB information for appropriate index and casts to use. The user should know whether the JsonParseObject is pointing to an array or an object and index appropriately. If they don't, we throw. How does having separate types resolve that issue? The user can still try to index a JsonParsedArray object with a string, no?